node/src/main/scala/com/wavesplatform/mining/BlockChallenger.scala

```diff
@@ -0,0 +1,277 @@
+import monix.execution.Scheduler
+
+import java.util.concurrent.ConcurrentHashMap
+import scala.concurrent.duration.*
+import scala.jdk.CollectionConverters.*
+
+trait BlockChallenger {
+  def challengeBlock(block: Block, ch: Channel, prevStateHash: ByteStr, blockReward: Option[Long]): Task[Unit]
+  def challengeMicroblock(md: MicroblockData, ch: Channel, prevStateHash: ByteStr, blockReward: Option[Long]): Task[Unit]
+  def pickBestAccount(accounts: Seq[(SeedKeyPair, Long)]): Either[GenericError, (SeedKeyPair, Long)]
+  def getChallengingAccounts(challengedMiner: Address): Either[ValidationError, Seq[(SeedKeyPair, Long)]]
+  def getProcessingTx(id: ByteStr): Option[Transaction]
+  def allProcessingTxs: Seq[Transaction]
+}
+
+object BlockChallenger {
+  val NoOp: BlockChallenger = new BlockChallenger {
+    override def challengeBlock(block: Block, ch: Channel, prevStateHash: ByteStr, blockReward: Option[Long]): Task[Unit]            = Task.unit
+    override def challengeMicroblock(md: MicroblockData, ch: Channel, prevStateHash: ByteStr, blockReward: Option[Long]): Task[Unit] = Task.unit
+    override def pickBestAccount(accounts: Seq[(SeedKeyPair, Long)]): Either[GenericError, (SeedKeyPair, Long)] =
+      Left(GenericError("There are no suitable accounts"))
+    override def getChallengingAccounts(challengedMiner: Address): Either[ValidationError, Seq[(SeedKeyPair, Long)]] = Right(Seq.empty)
+    override def getProcessingTx(id: ByteStr): Option[Transaction]                                                   = None
+    override def allProcessingTxs: Seq[Transaction]                                                                  = Seq.empty
+  }
+}
+
+class BlockChallengerImpl(
+    blockchainUpdater: BlockchainUpdater & Blockchain,
+    allChannels: ChannelGroup,
+    wallet: Wallet,
+    settings: WavesSettings,
+    timeService: Time,
+    pos: PoSSelector,
+    minerScheduler: Scheduler,
+    appendBlock: Block => Task[Either[ValidationError, Option[BigInt]]]
+) extends BlockChallenger
+    with ScorexLogging {
+
+  private val processingTxs: ConcurrentHashMap[ByteStr, Transaction] = new ConcurrentHashMap()
+
+  def challengeBlock(block: Block, ch: Channel, prevStateHash: ByteStr, blockReward: Option[Long]): Task[Unit] = {
+    log.debug(s"Challenging block $block")
+
+    (for {
+      _ <- EitherT.liftF(Task(processingTxs.putAll(block.transactionData.map(tx => tx.id() -> tx).toMap.asJava)))
+      challengingBlock <- EitherT(
+        createChallengingBlock(block, block.header.stateHash, block.signature, block.transactionData, prevStateHash, blockReward)
+      )
+      _ <- EitherT(appendBlock(challengingBlock).asyncBoundary)
+      _ <- EitherT.liftF[Task, ValidationError, Unit](Task(processingTxs.clear()))
+    } yield {
+      log.debug(s"Successfully challenged $block with $challengingBlock")
+      allChannels.broadcast(BlockForged(challengingBlock), Some(ch))
+    }).fold(
+      err => {
+        processingTxs.clear()
```

Comment:
There (and in `challengeMicroblock` too) could be `Task.bracket` to guarantee removing transactions from `processingTxs` even an error happened, e.g.:
```scala
  def challengeBlock(block: Block, ch: Channel, prevStateHash: ByteStr, blockReward: Option[Long]): Task[Unit] = {
    log.debug(s"Challenging block $block")

    Task(processingTxs.putAll(block.transactionData.map(tx => tx.id() -> tx).toMap.asJava)).bracket { _ =>
      (for {
        challengingBlock <- EitherT(
          createChallengingBlock(block, block.header.stateHash, block.signature, block.transactionData, prevStateHash, blockReward)
        )
        _ <- EitherT(appendBlock(challengingBlock).asyncBoundary)
      } yield {
        log.debug(s"Successfully challenged $block with $challengingBlock")
        allChannels.broadcast(BlockForged(challengingBlock), Some(ch))
      }).fold(
        err => log.debug(s"Could not challenge $block: $err"),
        identity
      )
    }(_ => Task(processingTxs.clear()))
  }
```

The next improvement is to save transactions and remove only them. But I'm not sure are concurrent challenges possible or not.

What do you think?

Reply:
Fixed
Concurrent challenges are not possible